File: packages/node-commons/src/public-key.ts
export interface PublicKeyPayload {
  environmentId: string;
  identityHost: string;
  version?: string;
  region?: "EU" | "US" | "APAC";
}

/**
 * Hosts the SDK is allowed to send the bearer token to. The identity host is
 * decoded from the (potentially attacker-influenced) public key, so it MUST be
 * validated against an allowlist before it is ever used as a `fetch` target —
 * otherwise a crafted `pk_` could point the SDK (and the token) at an internal
 * address (SSRF) or an attacker-controlled server (token exfiltration + auth
 * bypass).
 *
 * Additional hosts can be allowed via the `AUTHDOG_ALLOWED_IDENTITY_HOSTS`
 * environment variable (comma-separated hostnames), e.g. for self-hosted
 * identity servers.
 */
const DEFAULT_ALLOWED_HOST_SUFFIXES = ["authdog.com", "authdog.xyz"];

const isPrivateOrLoopbackHost = (hostname: string): boolean => {
  const h = hostname.toLowerCase();
  if (h === "localhost" || h.endsWith(".localhost")) return true;
  // IPv4 literals in private / loopback / link-local ranges
  if (/^127\./.test(h)) return true;
  if (/^10\./.test(h)) return true;
  if (/^192\.168\./.test(h)) return true;
  if (/^169\.254\./.test(h)) return true; // link-local (cloud metadata)
  if (/^172\.(1[6-9]|2\d|3[01])\./.test(h)) return true;
  // IPv6 loopback / unique-local
  if (h === "::1" || h === "[::1]") return true;
  if (h.startsWith("fc") || h.startsWith("fd") || h.startsWith("[fc") || h.startsWith("[fd"))
    return true;
  return false;
};

const getAllowedHostSuffixes = (): string[] => {
  const extra = (process.env.AUTHDOG_ALLOWED_IDENTITY_HOSTS ?? "")
    .split(",")
    .map((s) => s.trim().toLowerCase())
    .filter(Boolean);
  return [...DEFAULT_ALLOWED_HOST_SUFFIXES, ...extra];
};

/**
 * Validates that an identity host is safe to send credentials to: it must be a
 * well-formed `https:` URL whose hostname matches the allowlist and is not a
 * private/loopback address. Throws otherwise.
 */
export const assertTrustedIdentityHost = (identityHost: string): string => {
  let url: URL;
  try {
    url = new URL(identityHost);
  } catch {
    throw new Error("Invalid identity host");
  }

  if (url.protocol !== "https:") {
    throw new Error("Identity host must use https");
  }

  const hostname = url.hostname.toLowerCase();

  if (isPrivateOrLoopbackHost(hostname)) {
    throw new Error("Untrusted identity host");
  }

  const allowed = getAllowedHostSuffixes().some(
    (suffix) => hostname === suffix || hostname.endsWith(`.${suffix}`),
  );

  if (!allowed) {
    throw new Error("Untrusted identity host");
  }

  // Normalize: strip any trailing slash so callers can safely template URLs.
  return identityHost.replace(/\/+$/, "");
};

/**
 * Decodes and fully validates an Authdog public key (`pk_…`). This is the single
 * source of truth for parsing public keys — every framework SDK should use it
 * rather than re-implementing base64/JSON decoding.
 */
export const validateAndParsePublicKey = (publicKey: string): PublicKeyPayload => {
  if (!publicKey) {
    throw new Error("Public key is not defined");
  }

  if (!publicKey.startsWith("pk_")) {
    throw new Error("Invalid public key");
  }

  let payload: unknown;
  try {
    payload = JSON.parse(
      Buffer.from(publicKey.replace("pk_", ""), "base64").toString("utf-8"),
    );
  } catch {
    throw new Error("Failed to parse public key");
  }

  if (typeof payload !== "object" || payload === null) {
    throw new Error("Invalid public key payload");
  }

  const { environmentId, identityHost } = payload as Record<string, unknown>;

  if (typeof environmentId !== "string" || environmentId.length === 0) {
    throw new Error("Invalid public key: missing environmentId");
  }

  if (typeof identityHost !== "string" || identityHost.length === 0) {
    throw new Error("Invalid public key: missing identityHost");
  }

  // Reject any public key pointing at an untrusted identity host.
  const safeIdentityHost = assertTrustedIdentityHost(identityHost);

  return { ...(payload as PublicKeyPayload), identityHost: safeIdentityHost };
};

/**
 * @deprecated Use {@link validateAndParsePublicKey}, which additionally
 * validates the decoded payload and identity host. Kept as an alias for
 * backwards compatibility.
 */
export const getPublicKeyPayload = validateAndParsePublicKey;

File: packages/python/src/authdog/public_key.py
"""Public-key decoding and trusted identity-host validation.

This is the single source of truth for parsing Authdog public keys (``pk_…``)
in Python — it mirrors ``@authdog/node-commons`` so every framework SDK shares
the same base64/JSON decoding and the same SSRF / token-exfiltration guard.
"""

from __future__ import annotations

import base64
import binascii
import ipaddress
import json
import os
import re
from dataclasses import dataclass, field
from typing import Any
from urllib.parse import urlsplit

__all__ = [
    "PublicKeyPayload",
    "PublicKeyError",
    "assert_trusted_identity_host",
    "validate_and_parse_public_key",
]


class PublicKeyError(ValueError):
    """Raised when a public key (or its identity host) is invalid or untrusted."""


@dataclass(frozen=True)
class PublicKeyPayload:
    environment_id: str
    identity_host: str
    version: str | None = None
    region: str | None = None
    extra: dict[str, Any] = field(default_factory=dict)


# Hosts the SDK is allowed to send the bearer token to. The identity host is
# decoded from the (potentially attacker-influenced) public key, so it MUST be
# validated against an allowlist before it is ever used as a request target —
# otherwise a crafted ``pk_`` could point the SDK (and the token) at an internal
# address (SSRF) or an attacker-controlled server (token exfiltration).
DEFAULT_ALLOWED_HOST_SUFFIXES = ("authdog.com", "authdog.xyz")


def _is_private_or_loopback_host(hostname: str) -> bool:
    h = hostname.lower().strip("[]")
    if h == "localhost" or h.endswith(".localhost"):
        return True
    try:
        ip = ipaddress.ip_address(h)
    except ValueError:
        return False
    return (
        ip.is_private
        or ip.is_loopback
        or ip.is_link_local
        or ip.is_reserved
        or ip.is_unspecified
    )


def _allowed_host_suffixes() -> list[str]:
    extra = [
        s.strip().lower()
        for s in os.environ.get("AUTHDOG_ALLOWED_IDENTITY_HOSTS", "").split(",")
        if s.strip()
    ]
    return [*DEFAULT_ALLOWED_HOST_SUFFIXES, *extra]


def assert_trusted_identity_host(identity_host: str) -> str:
    """Validate that ``identity_host`` is safe to send credentials to.

    It must be a well-formed ``https:`` URL whose hostname matches the allowlist
    and is not a private/loopback address. Returns the host with any trailing
    slashes stripped; raises :class:`PublicKeyError` otherwise.
    """
    parts = urlsplit(identity_host)
    if parts.scheme != "https" or not parts.hostname:
        if parts.scheme and parts.scheme != "https":
            raise PublicKeyError("Identity host must use https")
        raise PublicKeyError("Invalid identity host")

    hostname = parts.hostname.lower()

    if _is_private_or_loopback_host(hostname):
        raise PublicKeyError("Untrusted identity host")

    allowed = any(
        hostname == suffix or hostname.endswith(f".{suffix}")
        for suffix in _allowed_host_suffixes()
    )
    if not allowed:
        raise PublicKeyError("Untrusted identity host")

    return identity_host.rstrip("/")


def validate_and_parse_public_key(public_key: str) -> PublicKeyPayload:
    """Decode and fully validate an Authdog public key (``pk_…``)."""
    if not public_key:
        raise PublicKeyError("Public key is not defined")

    if not public_key.startswith("pk_"):
        raise PublicKeyError("Invalid public key")

    raw = public_key[len("pk_") :]
    try:
        # urlsafe + padding tolerant: pad to a multiple of 4.
        padded = raw + "=" * (-len(raw) % 4)
        decoded = base64.b64decode(padded).decode("utf-8")
        payload = json.loads(decoded)
    except (binascii.Error, ValueError, UnicodeDecodeError) as exc:
        raise PublicKeyError("Failed to parse public key") from exc

    if not isinstance(payload, dict):
        raise PublicKeyError("Invalid public key payload")

    environment_id = payload.get("environmentId")
    identity_host = payload.get("identityHost")

    if not isinstance(environment_id, str) or not environment_id:
        raise PublicKeyError("Invalid public key: missing environmentId")
    if not isinstance(identity_host, str) or not identity_host:
        raise PublicKeyError("Invalid public key: missing identityHost")

    safe_host = assert_trusted_identity_host(identity_host)

    known = {"environmentId", "identityHost", "version", "region"}
    return PublicKeyPayload(
        environment_id=environment_id,
        identity_host=safe_host,
        version=payload.get("version"),
        region=payload.get("region"),
        extra={k: v for k, v in payload.items() if k not in known},
    )
